### PR TITLE
[WIP] Basic support for pre-training tasks

### PR DESCRIPTION
**Pull request by Joao-L-S-Almeida**
This is an work in progress, but it already supports the execution of simple pre-training tasks, that means, we can instantiate a transformer encoder-decoder architecture from scratch and train it to learn how to reconstruct a dataset.
 It still requires:

- [x]  at least one masked loss function. 
- [] a review for documentation.
- []  a scalability test.

### CODE REVIEW

**Comment by Joao-L-S-Almeida** [comment]
WARNING. If this PR isn't approved (or rejected) within 7 days it will be approved as is.

**Comment by romeokienzler** [comment]
closing as requested by @Joao-L-S-Almeida (solving conflicts in melting pot)
